src/filters/waterShed.ts

```diff
@@ -0,0 +1,196 @@
+          ) {
+            points.splice(i, 1);
+          }
+          break;
+        case false:
+          if (
+            image.getValue(points[i].column, points[i].row, 0) <= fillMaxValue
+          ) {
+            points.splice(i, 1);
+          }
+          break;
+        default:
+          throw new Error('unexpected value');
+      }
+    }
+  }
+
+  const maskExpectedValue = isMinimum ? 0 : 1;
+
+  const data = new Int16Array(currentImage.size);
+  const width = currentImage.width;
+  const height = currentImage.height;
+  const toProcess = new PriorityQueue({
+    comparator: (a: PointWithIntensity, b: PointWithIntensity) =>
+      a.intensity - b.intensity,
+    strategy: PriorityQueue.BinaryHeapStrategy,
+  });
+  for (let i = 0; i < points.length; i++) {
+    const index = points[i].column + points[i].row * width;
+    data[index] = i + 1;
+    const intensity = currentImage.getValueByIndex(index, 0);
+    if (
+      (isMinimum && intensity <= fillMaxValue) ||
+      (!isMinimum && intensity >= fillMaxValue)
+    ) {
+      toProcess.queue({
+        column: points[i].column,
+        row: points[i].row,
+        intensity,
+      });
+    }
+  }
+  const dxs = [+1, 0, -1, 0, +1, +1, -1, -1];
+  const dys = [0, +1, 0, -1, +1, -1, +1, -1];
+  // Then we iterate through each points
+
+  while (toProcess.length > 0) {
+    const currentPoint = toProcess.dequeue();
+    const currentValueIndex = currentPoint.column + currentPoint.row * width;
+    for (let dir = 0; dir < 4; dir++) {
+      const newX = currentPoint.column + dxs[dir];
+      const newY = currentPoint.row + dys[dir];
+      if (newX >= 0 && newY >= 0 && newX < width && newY < height) {
+        const currentNeighbourIndex = newX + newY * width;
+        if (
+          !mask ||
+          mask.getBitByIndex(currentNeighbourIndex) === maskExpectedValue
+        ) {
+          const intensity = currentImage.getValueByIndex(
+            currentNeighbourIndex,
+            0,
+          );
+          if (
+            ((isMinimum && intensity <= fillMaxValue) ||
+              (!isMinimum && intensity >= fillMaxValue)) &&
+            data[currentNeighbourIndex] === 0
+          ) {
+            data[currentNeighbourIndex] = data[currentValueIndex];
+            toProcess.queue({
+              column: currentPoint.column + dxs[dir],
+              row: currentPoint.row + dys[dir],
+              intensity,
+            });
+          }
+        }
+      }
+    }
+  }
+  const result: number[][] = [];
+  for (let row = 0; row < image.height; row++) {
+    const rowValues = [];
+    for (let column = 0; column < image.width; column++) {
+      rowValues.push(data[column + row * image.width]);
+    }
+    result.push(rowValues);
+  }
+  let resultMask = createMask(result);
+  if (isMinimum) {
+    resultMask = resultMask.invert();
+  }
+  //console.log(resultMask.invert(), fromMask(resultMask).getRois());
+
+  return fromMask(resultMask);
```

Comment:
We need to return a RoiMapManager seems to me. What is a mask doing here ?

Reply:
It does return a RoiMapManager. The problem that I encountered was that without nbPositive and nbNegative the RoiMapManager will not be correct. The only certain way I found to calculate it is to create a mask which then could be transformed into a RoiMapManager with correct number of positive and negative values on the map

Reply:
But you should not go through a Mask. You need to directly create:

https://github.com/image-js/image-js-typescript/blob/24a6b6622ead70c2264322501d9cf95279733554/src/roi/RoiMapManager.ts#L43-L47

A code similar to:

https://github.com/image-js/image-js-typescript/blob/24a6b6622ead70c2264322501d9cf95279733554/src/roi/fromMask.ts#L176-L182

Reply:
I understand but I need to know the nbNegative and nbPositive need to give the number of positive and negative Rois. And I don't know how to compute those without a createMask function. For example, in this example there are 2 negative rois and 2 positive ones. And if in this case the number of negative rois equals to length of extrema array, the number of negative rois is harder to find. Since we look for RoiMap we can't just put 0 as a value there.
![CleanShot 2023-08-09 at 11 03 17](https://github.com/image-js/image-js-typescript/assets/101188881/9a9dac9f-caf3-4f31-b71e-2ace423311d2)